rust/loqui_client/src/connection_handler.rs

```diff
@@ -0,0 +1,306 @@
+            config,
+        }
+    }
+}
+
+impl<E: Encoder> Handler for ConnectionHandler<E> {
+    type InternalEvent = InternalEvent<E::Encoded, E::Decoded>;
+    existential type UpgradeFuture: Send + Future<Output = Result<TcpStream, Error>>;
+    existential type HandshakeFuture: Send
+        + Future<
+            Output = Result<(Ready, FramedReaderWriter), (Error, Option<FramedReaderWriter>)>,
+        >;
+    existential type HandleFrameFuture: Send + Future<Output = Result<Response, (Error, u32)>>;
+
+    fn max_payload_size(&self) -> ByteSize {
+        self.config.max_payload_size
+    }
+
+    fn upgrade(&self, tcp_stream: TcpStream) -> Self::UpgradeFuture {
+        let max_payload_size = self.max_payload_size();
+        async move {
+            let framed_socket = Framed::new(tcp_stream, Codec::new(max_payload_size));
+            let (mut writer, mut reader) = framed_socket.split();
+            writer = match await!(writer.send(UpgradeFrame::Request)) {
+                Ok(writer) => writer,
+                Err(_e) => return Err(LoquiError::TcpStreamClosed.into()),
+            };
+            match await!(reader.next()) {
+                Some(Ok(UpgradeFrame::Response)) => Ok(writer.reunite(reader)?.into_inner()),
+                Some(Ok(UpgradeFrame::Request)) => Err(LoquiError::UpgradeFailed.into()),
+                Some(Err(e)) => {
+                    error!("Upgrade failed. error={:?}", e);
+                    Err(LoquiError::UpgradeFailed.into())
+                }
+                None => Err(LoquiError::TcpStreamClosed.into()),
+            }
+        }
+    }
+
+    fn handshake(&mut self, mut reader_writer: FramedReaderWriter) -> Self::HandshakeFuture {
+        async move {
+            let hello = Self::make_hello();
+            reader_writer = match await!(reader_writer.write(hello)) {
+                Ok(read_writer) => read_writer,
+                Err(e) => return Err((e.into(), None)),
+            };
+
+            match await!(reader_writer.reader.next()) {
+                Some(Ok(frame)) => match Self::handle_handshake_frame(frame) {
+                    Ok(ready) => Ok((ready, reader_writer)),
+                    Err(e) => Err((e, Some(reader_writer))),
+                },
+                Some(Err(e)) => Err((e, Some(reader_writer))),
+                None => Err((LoquiError::TcpStreamClosed.into(), Some(reader_writer))),
+            }
+        }
+    }
+
+    fn handle_frame(
+        &mut self,
+        frame: DelegatedFrame,
+        transport_options: &TransportOptions,
+    ) -> Option<Self::HandleFrameFuture> {
+        match frame {
+            DelegatedFrame::Response(response) => {
+                self.handle_response(response, transport_options);
+                None
+            }
+            DelegatedFrame::Error(error) => {
+                self.handle_error(error);
+                None
+            }
+            DelegatedFrame::Push(_) => None,
```

Comment:
Why not err?

Reply:
That would return an error frame back to the client for a push. Do we want to be sending frames back on a push?

Reply:
Guess it's fine to do that.